### ISSUE TITLE
Unify #sum, #sum:, #sumNumbers: and #detectSum:

### ISSUE THREAD

**Issue by theseion** [guillep/pharo-core#2225]
Migrated from https://pharo.fogbugz.com/f/cases/17174/Unify-sum-sum-sumNumbers-and-detectSum
//-----

Hi guys,

Collection defines `#sum:`, `#detectSum:` and `#sumNumbers:`, all of which accomplish the same (in principal) but with subtle differences:

`#sum:`
- uses #inject:into: with #anyOne as the injected element and will thus fail for empty collections

`#detectSum:`
- uses “nextValue + sum” instead of “sum + nextValue” which makes it a lot slower when dealing with large numbers (primitive fails and number conversion is necessary)

`#sumNumbers:`
- same as #sum but doesn’t fail for empty collections. Only good for numbers though.

Benchmarks:
```smalltalk
[ 100 timesRepeat: [ (1 to: 1000000) sum: #yourself ] ] timeToRun 18062
[ 100 timesRepeat: [ (1 to: 1000000) detectSum: #yourself ] ] timeToRun 42391
[ 100 timesRepeat: [ (1 to: 1000000) sumNumbers: #yourself ] ] timeToRun 18096
```

Can we settle for a single implementation? Such as (modified from `#sumNumbers:`):

```smalltalk
newSum: aBlock
        ^ self
                inject: (self
                        ifEmpty: [ 0 ]
                        ifNotEmpty: [ self anyOne ])
                into: [ :sum :each |  sum + (aBlock value: each) ]
```

This implementation combines the best of the three implementations I think. Benchmark:

```smalltalk
[ 100 timesRepeat: [ (1 to: 1000000) newSum: #yourself ] ] timeToRun 17955
```

BTW, there is also the message `#sum`, which suffers from the same problem as `#sum:` (the implementation is basically a copy). `#sum` could be implemented as

```smalltalk
sum
        ^ self sum: [ :x | x ]
```

As for the name for the new method, I suggest `#sum:`.

**Comment by theseion** [comment]
(Thierry Goubier)
//-----

Interesting results...

replacing the `#yourself` in `newSum:` cuts the runtime by a factor of two.
using an `[:x | x ]` block cost 10% compared to a direct version of sum.

So the fastest sum is:

```smalltalk
sum
    ^ self
        inject:
            (self ifEmpty: [ ^ 0 ] ifNotEmpty: [ self anyOne ])
        into: [ :sum :each | sum + each ]
```

And

```smalltalk
sum
    ^ self sum: [:x | x ]
```

is 10% slower.

And

```smalltalk
sum
    ^ self sum: #yourself
```

is 100% slower (i.e. x2)

It may be wise to avoid using `#yourself` for a block.

**Comment by theseion** [comment]
(Peter Uhnak)
//-----

The differences between `#sum:` and `#sumNumbers:` are well described in
the respective method comments, but I agree that this may be a problem
as most people will go for `#sum:` even for basic numeric collections.
(And I personally got bit by this.)

So maybe `#sum:` could adopt `#sumNumbers:` implementation, and the current
`#sum:` implementation would be moved to `#domainSum:` (or some other
appropriate name...)?

As for `#detectSum:` I think this method has confusing name and is
functionally duplicate to #sum:.

The swapped arguments there may make sense when you overload `#+`, however
the same thing can be achieved from the provided block ...

**Comment by theseion** [comment]
(Sven Van Caekenberghe)
//-----

I am all for a cleanup, the current situation is confusing.
The basic `#sum` should be fast AND work for empty collections with 0 as starting element.
I know why the `#anyOne` is used, and that use case should be preserved, but it is less common IMHO.

**Comment by theseion** [comment]
(Ben Coman)
//-----

>I know why the `#anyOne` is used, and that use case should be preserved, but it is less common IMHO.

I'm curious to be learn why?

**Comment by theseion** [comment]
(Thierry Goubier)
//-----

Note that the sum with `#anyOne` is wrong.

```smalltalk
#(1 2 3 4 5) inject: #(1 2 3 4 5) anyOne into: [ :sum :each | sum + each ]
```

returns 16 instead of 15.

**Comment by theseion** [comment]
(Sven Van Caekenberghe)
//-----

>I'm curious to be learn why?

How many times have you summed a collection of things, where (1) you could not start from the number 0 and (2) the operation was `#+` ? Me, never, and if I would need that, I would do an `#inject:into:` myself. Any counter examples that occur a lot ?

**Comment by theseion** [comment]
(Sven Van Caekenberghe)
//-----

> Note that the sum with #anyOne is wrong.
>
>```smalltalk
>#(1 2 3 4 5) inject: #(1 2 3 4 5) anyOne into: [ :sum :each | sum + each ]
>```
>
> returns 16 instead of 15.

You have subtract the element you picked with `#anyOne` !

**Comment by theseion** [comment]
(Nicolai Hess)
//-----

> > Note that the sum with #anyOne is wrong.
> >
> > ```smalltalk
> > #(1 2 3 4 5) inject: #(1 2 3 4 5) anyOne into: [ :sum :each | sum + each ]
> >```
> >
> > returns 16 instead of 15.

>You have subtract the element you picked with `#anyOne` !

I think he means the suggested implementation for `newSum` is wrong, just because of this reason.

**Comment by theseion** [comment]
(Nicolai Hess)
//-----

> How many times have you summed a collection of things, where (1) you could not start from the number 0 and (2) the operation was `#+` ? Me, never, and if I would need that, I would do an `#inject:into:` myself. Any counter examples that occur a lot ?

wokring with things like `Color`

```smalltalk
(Color wheel:10) sum "works"
(Color wheel:10) newSum:#yourself " would not work"
```

**Comment by theseion** [comment]
(Thierry Goubier)
//-----

> I think he means the suggested implementation for `newSum` is wrong, just because of this reason.

Exactly ;)

**Comment by theseion** [comment]
(Sven Van Caekenberghe)
//-----

The basic question for me is, what should

```smalltalk
 #() sum
```

return. Right now, it is an error, I would very much like that for this common case the result would be 0. There is a lot of power (easy of use) in a unary selector, we should not destroy that with semantics that force a test before using it.

If we agree on that, we could change the #sum to

```smalltalk
| sum sample |
self isEmpty ifTrue: [ ^ 0 ].
sample := self anyOne.
sum := self inject: sample into: [ :accum :each | accum + each ].
^ sum - sample
```

And the same for `#sum:`

As for `#detectSum:` that should be removed I think.

**Comment by theseion** [comment]
(Thierry Goubier)
//-----

>```smalltalk
>  #() sum
>```

> return. Right now, it is an error, I would very much like that for this common case the result would be 0. There is a lot of power (easy of use) in a unary selector, we should not destroy that with semantics that force a test before using it.

Yes, this is just that a 0 value is returned for an empty collection, and, if you're expecting a sum of Colors, then this is a bit strange.

Returning an error on a sum for an empty collection is maybe an overall cleaner concept

Which can be expressed as:

I prefer this code:
```smalltalk
aCollectionOfColors isEmpty
    ifTrue: [ sumOfColors := aDefaultColor ]
    ifFalse: [ sumOfColors := aCollectionOfColors sum ]
```
To this code:
```smalltalk
sumOfColors := aCollectionOfColors sum.
sumOfColors = 0
    ifTrue: [ sumOfColors := aDefaultColor ]
 ```

If we agree on that, we could change the #sum to
```smalltalk
| sum sample |
self isEmpty ifTrue: [ ^ 0 ].
sample := self anyOne.
sum := self inject: sample into: [ :accum :each | accum + each ].
^ sum - sample
```
Ok, now I really like this implementation. Elegant.

**Comment by theseion** [comment]
(Peter Uhnak)
//-----

> I prefer this code:
>```smalltalk
> aCollectionOfColors isEmpty
>    ifTrue: [ sumOfColors := aDefaultColor ]
>    ifFalse: [ sumOfColors := aCollectionOfColors sum ]
>```
> To this code:
>```smalltalk
> sumOfColors := aCollectionOfColors sum.
> sumOfColors = 0
>  ifTrue: [ sumOfColors := aDefaultColor ]
>```

or
```smalltalk
sumOfColors := (aCollectionOfColors ifEmpty: [ {aDefaultColor} ]) sum.
```

> If we agree on that, we could change the #sum to
>```smalltalk
> | sum sample |
> self isEmpty ifTrue: [ ^ 0 ].
> sample := self anyOne.
> sum := self inject: sample into: [ :accum :each | accum + each ].
> ^ sum - sample
>```

> Ok, now I really like this implementation. Elegant.

Why the temp variables?
```smalltalk
self ifEmpty: [ ^ 0 ].
^ self inject: self anyOne negated into: [ :sum :each | sum + each ]
```
But more importantly, if the empty/default is 0 then what is the reason for `#anyOne`?
Otherwise you are mixing things.
```smalltalk
self ifEmpty: [ ^ 0 ].
^ self inject: 0 into: [ :sum :each | sum + each ].
```

**Comment by theseion** [comment]
(Stephan Eggermont)
//-----

> The basic question for me is, what should
>```smalltalk
>  #() sum
>````
>
> return. Right now, it is an error, I would very much like that for this common case the result would be 0. There is a lot of power (easy of use) in a unary selector, we should not destroy that with semantics that force a test before using it.

I like the error, it aligns with most of our collection protocol.
It shows the need for `#sum:ifEmpty:` though

**Comment by theseion** [comment]
(Sven Van Caekenberghe)
//-----

> I like the error, it aligns with most of our collection protocol.

I hate the error, a lot ;-)

> It shows the need for `#sum:ifEmpty:` though

Yes, as long as `#() sum == 0 `

I want the simplest case to be simple, having a non-0 default is a special case IMHO

**Comment by theseion** [comment]
Well, I’m glad I've struck a nerve there :)

I agree that signalling an exception isn’t wrong, it’s just annoying because you have to do the check manually each time. I think I like `#sum:ifEmpty:`. When you look through the methods in Collection you’ll see `#sum`, `#sum:` and `#sum:ifEmpty:` together in one place and it will be clear how to use them.

> I hate the error, a lot ;-)

+1

> > It shows the need for `#sum:ifEmpty:` though

> Yes, as long as `#() sum == 0` 

+100

That allows for nice Smalltalk like `myNumbers sum isZero …`.

**Comment by theseion** [comment]
(Yuriy Tymchuk)
//-----

`#sum:ifEmpty:` looks nice

**Comment by theseion** [comment]
(Tudor Girba)
//-----

Hi,


> > It shows the need for #sum:ifEmpty: though

> Yes, as long as #() sum == 0 

That won’t work :).

> > I want the simplest case to be simple, having a non-0 default is a special case IMHO

That is why you have `sumNumbers:`. We could also add `Collection>>sumNumbers`.

We had this discussion at length before Pharo 4, and this is when we agreed to add `sumNumbers:` and let `sum:` be generic (like the name says it should be) and not assume that it should work with Numbers.

**Comment by theseion** [comment]
(Yuriy Tymchuk)
//-----

> That won’t work :).

But wouldn’t it be nice to have `#sum:ifEmpty:`? I think that this is a nice way to describe summing. Otherwise I have to do `ifEmpty: [ something ] ifNotEmpty: [ :col | col sum: … ]`

**Comment by theseion** [comment]
(Sven Van Caekenberghe)
//-----

Doru,

> That won’t work :).

Why ? Please explain.

> That is why you have `sumNumbers:`. We could also add `Collection>>sumNumbers`.
>
> We had this discussion at length before Pharo 4, and this is when we agreed to add `sumNumbers:` and let `sum:` be generic (like the name says it should be) and not assume that it should work with Numbers.

It is not about numbers or not, you are still using `#+` in your generic case, that is numeric in my book. It is about the zero element and how to deal with an empty collection.

The current solution, which was indeed recently added, leaves the problem of what to do with an empty collection. There is no general solution, you have to specify a zero element.

But for most people in most cases that will be effectively 0, so lets make the unary `#sum` respect that. (And unary `#sum` will also work for non-empty non-zero based objects). The less common case can then use the longer message.

**Comment by theseion** [comment]
(Tudor Girba)
//-----

Hi,

> But wouldn’t it be nice to have `#sum:ifEmpty:`? I think that this is a nice way to describe summing. Otherwise I have to do `ifEmpty: [ something ] ifNotEmpty: [ :col | col sum: … ]`

Sure. That would be fine, and would leave the meaning of `sum` to work with non-numbers objects.

But, of course, you can also use `sumNumbers:` which starts from 0 not from `anyOne` and you will need no `ifEmpty:`.

**Comment by theseion** [comment]
(Esteban Maringolo)
//-----

I don't want to be heretic (or too orthodox), but why not to delegate
this behavior to other class (an iterator maybe?).

It's too tempting adding these convenience methods to Collection
and/or subclasses, but anything that requires an explicit protocol of
its elements is wrong, IMO.

something like `aCollection arithmetic sum: [...]` or.... `aCollection
arithmetic avg`.

My two cents for this.

**Comment by theseion** [comment]
Doru,
You’re missing the point: `#anyOne` *fails* for empty collections.


> I don't want to be heretic (or too orthodox), but why not to delegate
> this behavior to other class (an iterator maybe?).
>
> It's too tempting adding these convenience methods to Collection
> and/or subclasses, but anything that requires an explicit protocol of
> its elements is wrong, IMO.
>
> something like `aCollection arithmetic sum: [...]` or.... `aCollection arithmetic avg`.


Interesting thought!

**Comment by theseion** [comment]
(Ben Coman)
//-----

https://en.wikipedia.org/wiki/Empty_sum

**Comment by theseion** [comment]
(Chris Cunningham)
//-----

Hmm.

Interesting. On my machine (which is apparently really slow right now), I get:
```smalltalk
  [ 100 timesRepeat: [ (1 to: 1000000) detectSum: [:x|x] ] ] timeToRun  "104874"
  [ 100 timesRepeat: [ (1 to: 1000000) detectSumSwitch: [:x|x] ] ] timeToRun "44369"
```
(this second one is where I add have switched the order - `sum + nextValue` - for speed.  So, a little over twice as fast).

If I change `SmallInteger `to be a bit smarter, I get:
```smalltalk
  [ 100 timesRepeat: [ (1 to: 1000000) detectSum: [:x|x] ] ] timeToRun "45257"
```
In other words just marginally slower than the optimized method above. The benefit here is that future folks won't have to remember which way to do their arithmetic to make things fast.

Smarter `+` (in `SmallInteger`):
```smalltalk
+ aNumber 
 "Primitive. Add the receiver to the argument and answer with the result
    if it is a SmallInteger. Fail if the argument or the result is not a
    SmallInteger  Essential  No Lookup. See Object documentation whatIsAPrimitive." 

    <primitive: 1>
    ^aNumber class = SmallInteger
        ifFalse: [aNumber + self]
        ifTrue: [super + aNumber]
```

Of course, this wants to be thoroughly tested before use - certain additions may now work in ways not expected before (such as adding an array to a small integer - you may get odd results).  Still...

**Comment by theseion** [comment]
(Tudor Girba)
//-----

Hi,

> Doru,
> You’re missing the point: #anyOne *fails* for empty collections.

I am not missing the point at all. I am saying that if you want `sum:` to be generic, it cannot assume a specific Zero object.

And `sum: `should be generic because of its name.


> Interesting thought!

+100

**Comment by theseion** [comment]
(Jeff Rick)
//-----

Lovely little discussion which would be great for beginning OO programmers to see. The question of whether to return 0 for an empty collection is interesting. While other things can be summed, we tend to think of sums returning numbers. In that sense, 0 is a great default return. On the other hand, having a specific object for a generic sum function is problematic (+1). There is a real tension and that would be a great discussion for people to have. Either way, the need for `sumIfEmpty:` is obvious.

**Comment by theseion** [comment]
(Tudor Girba)
//-----

> It is not about numbers or not, you are still using `#+` in your generic case, that is numeric in my book. It is about the zero element and how to deal with an empty collection.
>
> The current solution, which was indeed recently added, leaves the problem of what to do with an empty collection. There is no general solution, you have to specify a zero element.

Exactly.

> But for most people in most cases that will be effectively 0, so lets make the unary #sum respect that. (And unary `#sum `will also work for non-empty non-zero based objects). The less common case can then use the longer message.

You said it correctly: “most” :). That is why, a generic meaning should be left generic and provide a solution that will work with “all” cases :).

It’s a matter of choice, but `sum:` has a too generic meaning to have it confined to numbers.

**Comment by theseion** [comment]
(Sven Van Caekenberghe)
//-----

> You said it correctly: “most” :). That is why, a generic meaning should be left generic and provide a solution that will work with “all” cases :).
>
> It’s a matter of choice, `but sum:` has a too generic meaning to have it confined to numbers.

I think we all mean the same thing and understand each other's point.

But the first meaning of `sum` in Wikipedia leads to https://en.wikipedia.org/wiki/Summation which talks about numbers. I studied partly Mathematics, so I know there are different numeric systems and abstractions. My point is: the common case is adding (simple) numbers, the special case is the more general one, hence the best selector, `#sum`, should work for the common case, including for the empty one.

Consider also Ben's link:

https://en.wikipedia.org/wiki/Empty_sum

Hence my suggestion to change `#sum` to
```smalltalk
| sum sample |
self isEmpty ifTrue: [ ^ 0 ].
sample := self anyOne.
sum := self inject: sample into: [ :accum :each | accum + each ].
^ sum - sample
```
And the same for `#sum:`

Note that the above *will* work for all like today, just not if they are empty. Now, of course we will need the `ifEmpty:` variants for what I consider the less common cases. 

Can I also remark that both `#inject:into:` and `#fold:` are perfectly fine for more involved cases. We have a lot of API ;-)

**Comment by theseion** [comment]
(Ben Coman)
//-----

> And `sum:` should be generic because of its name.

I am missing understanding the other use cases.  Can you describe
further the generic nature of `#sum` & `#sum:` ?  I would have thought by
default they only applied to numbers.

**Comment by theseion** [comment]
(Nicolai Hess)
//-----

> I am missing understanding the other use cases.  Can you describe
> further the generic nature of `#sum` & `#sum:` ?  I would have thought by
> default they only applied to numbers.

`sum` can be applied to anything that supports `#+`, not only numbers
`sum: `can be applied to any collection with a block that return some object that supports `#+`

therefore you can not assume 0 (<- a number) is a proper initial value therefore you *need* to work with `#anyOne`
and as you can not assume a proper initial value, you can not assume a default value for empty collections
-> it should throw an error. If you (the caller of the function) knows what to do with an empty collection you have
to check, or call `inject:into:` directly, with a proper initial value.

**Comment by theseion** [comment]
(Sven Van Caekenberghe)
//-----

> `sum` can be applied to anything that supports `#+`, not only numbers
> `sum:` can be applied to any collection with a block that return some object that supports `#+`

> therefore you can not assume 0 (<- a number) is a proper initial value therefore you *need* to work with `#anyOne`
> and as you can not assume a proper initial value, you can not assume a default value for empty collections
> -> it should throw an error. If you (the caller of the function) knows what to do with an empty collection you have
> to check, or call inject:into: directly, with a proper initial value.

I am sorry but I am getting really tired of this, you should read what is being said.

I am not suggesting to stop using `#anyOne` because I like why it is there and what it can do.

The change I want is what happens with an empty collection when using the simplest selector, `#sum.

I do not want to say to some collection of numbers #sumIfEmpty: [0], because summing numbers starting from zero is the most common case and everybody expects that, hence the unary selector fits.

I want the less common cases to use the more complicated API, as in some collection of colors 
```smalltalk
#sumIfEmpty: [ Color black ]
```

In my book that is common sense API design.

Doing that I do no take anything away, because today you already have to make sure the collection is not empty.

The only change would be in the error behaviour. I think that is a reasonable price to pay. Instead of having `#anyOne` fail, it will say that `#+` cannot add 0 to some object, and in a comment we can point to the alternative API.

http://izquotes.com/quote/242740 right ?

**Comment by theseion** [comment]
(Sven Van Caekenberghe)
//-----

> The only change would be in the error behaviour. I think that is a reasonable price to pay. Instead of having #anyOne fail, it will say that `#+` cannot add 0 to some object, and in a comment we can point to the alternative API.

Make that: instead of having `#anyOne` fail (which is pretty obscure) it will return 0 which will then probably fail.

**Comment by theseion** [comment]
(Ben Coman)
//-----

> `sum` can be applied to anything that supports `#+`, not only numbers
> `sum:` can be applied to any collection with a block that return some object that supports `#+`

To me this is a mis-application of polymorphism, that just because
something responds to `#+` it should be summable. We have overloaded the
semantics of  `#+`  to mean both numeric addition and
concatenation/membership, but technically "summation" relates only to
numeric addition.

https://www.google.com.au/search?q=define+sum&oq=define+sum
https://www.google.com.au/search?q=define+concatenate&oq=define+concatenate

For example...

- `KMModifier` implements  `#+`  so what is the expected semantic of 
  ```smalltalk
  {KMModifier shift . KMModifier meta} sum
  ```
  ? To me this is more of a concatenation/join/union facility rather than
numeric addition.  (btw, that expression actually fails since
`KMModifier` does not understand minus `#-` ) .

- `String` implements `#+`  and 
  ```smalltalk
  { '1' . '2' } sum "--> '3'"
  ```
  ,  so actually its doing numeric addition.  However
  ```smalltalk
  { 'a' . 'b' } sum
  ```
  produces an error.

  So actually there seem some existing problems with summing
non-numerics.  What examples work?

- Trait classes implement both  `#+`  and  `#-` , but the semantic seems
more to do with membership than numeric addition.  I don't how how to
produce an example of using #sum against traits.

- Points are summable
  ```smalltalk
  { 2@2 . 3@3 } "--> 5@5"
  ```
  But then 
  ```smalltalk
  2@2 + 1  "--> 3@3"
  ```
  ,  so  `{} sum`  returning  0  would seem to not
cause any error in this case.

**Comment by theseion** [comment]
(Chris Cunningham)
//-----

> -Points are summable
>   ```smalltalk
>   { 2@2 . 3@3 } "--> 5@5"
>   ```
>   But then 
>   ```smalltalk
>   2@2 + 1  "--> 3@3"
>   ```
>   ,  so  `{} sum`  returning  0  would seem to not
> cause any error in this case.


but points aren't commutative:
```smalltalk
2@2 + 1 " = 3@3"
1 + 2@2 " = 3@2"
````
Of course, 0 wouldn't be an issue, unless you wanted to access x or y!

**Comment by theseion** [comment]
(Nicolai Hess)
//-----

> To me this is a mis-application of polymorphism, that just because
> something responds to `#+` it should be summable. We have overloaded the
> semantics of  `#+`  to mean both numeric addition and
> concatenation/membership, but technically "summation" relates only to
> numeric addition.

I didn't wanted to argue for or against any change. I just wanted to clarify
that there are situations in which a generic `sum`/`sum:` that throws an error
on empty collections and don't assume a null value makes sense.

> I am sorry but I am getting really tired of this, you should read what is being said.

do that change, I am not against it. Ben just asked for an example and I thought it would be helpful.

**Comment by theseion** [comment]
(Ben Coman)
//-----

> do that change, I am not against it. Ben just asked for an example and I
> thought it would be helpful.

It was helpful :)  It evolved my thinking.  Now thinking further, I
wonder how returning  0  will work with applications using units like
Aconcagua, and if it would over-complicate things to do something
like...
```amalltalk
  Collection>>sum
      | sum sample |
      self isEmpty ifTrue: [ ^ ArithmeticZero ].
      sample := self anyOne.
      sum := self inject: sample into: [ :accum :each | accum + each ].
      ^ sum - sample
```
```smalltalk
  ArithmeticZero class >> + anObject
      ^anObject
```

**Comment by theseion** [comment]
(Ben Coman)
//-----

> but points aren't commutative:
>```smalltalk
> 2@2 + 1 " = 3@3"
> 1 + 2@2 " = 3@2"
>```
>
> Of course, 0 wouldn't be an issue, unless you wanted to access x or y!


whoops?
```smalltalk
1 + (2@2)  = 3@3
```

**Comment by theseion** [comment]
(Chris Cunningham)
//-----

> whoops?
>```smalltalk
> 1 + (2@2)  = 3@3
>```

Yeah.  Forgot the parenthesis.  `1 + 2@2` is the same as `(1 + 2) @ 2`.
My eye sees the closeness of `2@2` as a point - not as a binary message (which it obviously is). 

So, never mind.

**Comment by theseion** [comment]
> To me this is a mis-application of polymorphism, that just because
> something responds to `#+` it should be summable. We have overloaded the
> semantics of  `#+`  to mean both numeric addition and
> concatenation/membership, but technically "summation" relates only to
> numeric addition.

I agree. If I want to sum something other than a collection of numbers I’ll happily provide the block for that summation (I wouldn’t even expect strings to be summable to be honest. Why would you have a collection of numbers that are strings…? And if you really have one, you should convert it explicitly so that another person reading your code knows what’s going on.)


> It was helpful :)  It evolved my thinking.  Now thinking further, I
> wonder how returning  0  will work with applications using units like
> Aconcagua, and if it would over-complicate things to do something
> like...
>```smalltalk
>  Collection>>sum
>      | sum sample |
>      self isEmpty ifTrue: [ ^ ArithmeticZero ].
>      sample := self anyOne.
>      sum := self inject: sample into: [ :accum :each | accum + each ].
>      ^ sum - sample
>```
>```smalltalk
>  ArithmeticZero class >> + anObject
>      ^anObject
>```

While I think you might be on to something, I think we should take small steps. I’d be happy already if we can just get rid of one superfluous method and provide a better API without starting to think about the deeper semantics.
Also, I think there’s a reason why Aconcagua is an external package: developers usually are happy to work with numbers (which are already objects in Smalltalk anyway) and they’re fast.

**Comment by theseion** [comment]
Instead of copying this (endless) thread here, I'm linking to it: http://thread.gmane.org/gmane.comp.lang.smalltalk.pharo.devel/114899/focus=115734

**Comment by theseion** [comment]
We're now in feature freeze. Moving this to Pharo 6.0.

**Comment by theseion** [comment]
(Marcus Denker)
//-----
what is the status of this?

**Comment by theseion** [comment]
Good question. I never heard back from anyone after the initial (heated) discussion. I hope to tackle all my open issues in the next couple of weeks.

**Comment by theseion** [comment]
(Andrew Black)
//-----

I've just read through this rather old thread, and want to add an observation. Starting the calculation of `sum` with `anyOne` and then subtracting `anyOne` at the end works only if 
 - `+` is commutative
 ` the elements support `-` as well as `+`, and
 - `+ and `-` are inverses.

A simple case where this won't work is strings with `+` for concatenation.  

It's more general to simply start the `sum` with collection first, and then add in elements `2 .. collection size`.  And that will probably be faster too, since `Number>>#to:do:` is typically faster than `collections>>#do:`

**Comment by theseion** [comment]
(Cédrik Béler)
//-----

read parts of the thread. I agree sum cannot return 0 as default. But they're should be at least a proper error (like `#sum` cannot be applied to an empty collection) and maybe an alternative sum like `Collection>>sumDefault:`  (don't like the selector name though). 

But then (this was probably discussed I guess), we could have a generic way (each summable collection item defining a default nul value 
```smalltalk
SmallInteger>>defaultZero
  ^0
```
```smalltalk
Point>>defaultZero
  ^0@0
```
 ...      just thinking loud

**Comment by theseion** [comment]
This is clearly a complicated issue.

Stef: maybe this could be the topic of a seminar project or even a bachelor's thesis?

For reference, Stefan Marr published a paper with a review of collection API designs: http://stefan-marr.de/papers/px-marr-daloze-few-versatile-vs-many-specialized-collections/

**Comment by girba** [comment]
For reference, the previous discussions can be found here:
https://pharo.fogbugz.com/f/cases/12182/Collection-sumNumbers-should-provide-an-alternative-to-failing-sum
https://pharo.fogbugz.com/f/cases/17174/Unify-sum-sum-sumNumbers-and-detectSum

I also remember the conversations from the mailing list that listed strong reasons for not implying `0` for the generic `sum` or `sum:`.

**Comment by svenvc** [comment]
I also remember ;-)

I remain on the position that #() sum should be zero.

Why ?

- https://en.wikipedia.org/wiki/Empty_sum
- 9 out 10 cases, you will be summing numbers
- unary selectors are special for being so convenient and easy to use
- the most common case should be the simplest and least surprising

The less common case, summing non-numbers that are not compatible with 0 as additive/zero element can be covered with #sumIfEmpty: using the less elegant selector.

In code:
```
widths sum. "Everybody understand this will be zero if empty"
colors sumIfEmpty: [ Color black ].
```

**Comment by girba** [comment]
In the debate from 2013, I was advocating for `0` to be the default return value when the collection is empty. People said it is better not to, and we introduced `sumNumbers`. After having worked with a couple of systems that relied on money, prices and various units, I was happy that `sum` was not `0` by default. I still like having distinct `sum` and `sumNumbers`, but I can understand the other perspective as well.

**Comment by theseion** [comment]
Still relevant.

**Comment by macta** [comment]
> In the debate from 2013, I was advocating for `0` to be the default return value when the collection is empty. People said it is better not to, and we introduced `sumNumbers`. After having worked with a couple of systems that relied on money, prices and various units, I was happy that `sum` was not `0` by default. I still like having distinct `sum` and `sumNumbers`, but I can understand the other perspective as well.

I just got hit by this one - this doesn't just impact sum, but things that use sum - like #average (as it calls sum, and not sumNumbers). If you take the average of a series of floats you get a rounding difference.

e.g. `
 {987.9504418944 . 815.2627636718801 . 1099.3898999037601  . 1021.6996069333198  . 1019.8750146478401 . 1084.5603759764 . 1008.2985131833999 . 1194.9564575200002 . 893.9680444336799 . 1032.85460449136 . 905.9324633786798 . 1024.2805590819598 . 784.5488305664002 . 957.3522631840398 . 1001.7526196294} average.
`
Gives:
`988.8454972331014`

and not: 
`988.8454972331015`

This is annoying for basic math type stuff - and the above example comes from getting data from an historical weather api, and the examples in Python with tests don't convert cleanly to Pharo without delving into the guts of what should be easy stuff - I just wanted the #average and didn't want to know about this stuff.

Looking at other things that use #sum (a lot of the graphics stuff does), I wonder how many subtle rounding errors are giving offset circles etc.

**Comment by theseion** [comment]
Thanks Tim!

**Comment by gcorriga** [comment]
This can be closed following the changes in PRs #6453 and #8299
